crates/doc/src/inference.rs

```diff
@@ -899,6 +909,9 @@ impl Shape {
         let default = union_option(lhs.default, rhs.default);
         let secret = union_option(lhs.secret, rhs.secret);
 
+        let mut annotations = lhs.annotations;
+        annotations.extend(rhs.annotations.into_iter());
```

Comment:
Yeah, that's actually what I started out doing, but I realized it doesn't actually work for annotations. Concretely, the problem is due to [this code](https://github.com/estuary/flow/blob/6183db1396e68e595af2f4d71779a1f6b42ea0ad/crates/doc/src/inference.rs#L873), which intersects the shapes of `oneOf` and `allOf` with the base shape. Consider a schema like:

```
x-outer-annotation: 1
oneOf:
  - x-some-annotation: true
```

If you intersect the top-level annotations with the `oneOf`, then the result will always be empty.

Less concretely, I think semantically the `annotations` of a shape are all the annotations that _might_ apply at a given location. If you want only the annotations that _do_ apply, then you need to actually validate a document. So like in [this example](https://github.com/estuary/flow/blob/6183db1396e68e595af2f4d71779a1f6b42ea0ad/crates/doc/src/inference.rs#LL2448C28-L2448C28) we should expect the Shape to include [all four annotations](https://github.com/estuary/flow/blob/6183db1396e68e595af2f4d71779a1f6b42ea0ad/crates/doc/src/snapshots/doc__inference__test__annotation_collection.snap#L44)

Reply:
I'm not sure I'm following the problem here.

`one_of`, in the code you linked, will be the single sub-schema. It will be intersected with the parent schema. That intersection actually becomes `child.annotations.extend(parent.annotations)`, which is what you want, right?

Reply:
> I think semantically the annotations of a shape are all the annotations that might apply at a given location. 

That's not how `Shape` works for other annotations, like `title`, `format`, and so on. 
`Shape` tells us only what we know is _always_ true about a document location.

Let's talk through over VC?

Reply:
Caught up on VC about this, and actually the annotations in a Shape should only include the annotations that are _guaranteed_ to apply at a particular location. Given that, I'm changing this back to an intersection.